File: 6_tree/多叉树/四叉树/427. 建立四叉树.ts
class Node {
  val: boolean // val：储存叶子结点所代表的区域的值。1 对应 True，0 对应 False
  isLeaf: boolean // 当这个节点是一个叶子结点时为 True，如果它有 4 个子节点则为 False
  topLeft: Node | null
  topRight: Node | null
  bottomLeft: Node | null
  bottomRight: Node | null
  constructor(
    val: boolean,
    isLeaf: boolean,
    topLeft: Node | null,
    topRight: Node | null,
    bottomLeft: Node | null,
    bottomRight: Node | null
  ) {
    this.val = val
    this.isLeaf = isLeaf
    this.topLeft = topLeft
    this.topRight = topRight
    this.bottomLeft = bottomLeft
    this.bottomRight = bottomRight
  }
}

function construct(grid: (0 | 1)[][]): Node | null {
  return makeTree(0, 0, grid.length)

  function makeTree(x: number, y: number, edgeLength: number): Node | null {
    const val = getMatrixValue(x, y, edgeLength)
    let node: Node

    if (val === 2) {
      const half = ~~(edgeLength / 2)
      node = new Node(
        true,
        false,
        makeTree(x, y, half),
        makeTree(x, y + half, half),
        makeTree(x + half, y, half),
        makeTree(x + half, y + half, half)
      )
    } else {
      node = new Node(val === 1 ? true : false, true, null, null, null, null)
    }

    return node
  }

  function getMatrixValue(x: number, y: number, edgeLength: number): 0 | 1 | 2 {
    const val = grid[x][y]

    for (let i = x; i < x + edgeLength; i++) {
      for (let j = y; j < y + edgeLength; j++) {
        if (grid[i][j] !== val) return 2
      }
    }

    return val
  }
}

console.log(
  construct([
    [0, 1],
    [1, 0],
  ])
)
// 输出：[[0,1],[1,0],[1,1],[1,1],[1,0]]   [0,1] 表示虚拟根节点不是叶子 值为1

// 判断当前矩阵是否全为1或全为0，如果是则创建四叉树叶子节点，
// 否则创建四叉树非叶子节点，4个子分别是将矩阵分为4份构成的子树
// 时间复杂度：O(n^4)
// 空间复杂度：O(logn)

export { construct }

File: 6_tree/多叉树/四叉树/427. 建立四叉树.py
# Definition for a QuadTree node.
from typing import List


class Node:
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val  # 储存叶子结点所代表的区域的值。1 对应 True，0 对应 False；
        self.isLeaf = isLeaf  # 当这个节点是一个叶子结点时为 True，如果它有 4 个子节点则为 False
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight


class PreSumMatrix:
    """二维前缀和模板(矩阵不可变)"""

    def __init__(self, A: List[List[int]]):
        m, n = len(A), len(A[0])

        # 前缀和数组
        preSum = [[0] * (n + 1) for _ in range(m + 1)]
        for r in range(m):
            for c in range(n):
                preSum[r + 1][c + 1] = A[r][c] + preSum[r][c + 1] + preSum[r + 1][c] - preSum[r][c]
        self.preSum = preSum

    def sumRegion(self, r1: int, c1: int, r2: int, c2: int) -> int:
        """查询sum(A[r1:r2+1, c1:c2+1])的值::

        preSumMatrix.sumRegion(0, 0, 2, 2) # 左上角(0, 0)到右下角(2, 2)的值
        """
        return (
            self.preSum[r2 + 1][c2 + 1]
            - self.preSum[r2 + 1][c1]
            - self.preSum[r1][c2 + 1]
            + self.preSum[r1][c1]
        )


class Solution:
    def construct(self, grid: List[List[int]]) -> 'Node':
        def dfs(r1: int, c1: int, r2: int, c2: int) -> Node:
            """类似于克隆图"""
            val = preSumMatrix.sumRegion(r1, c1, r2 - 1, c2 - 1)

            # 叶子结点
            if val == 0:
                return Node(False, True, None, None, None, None)
            elif val == (r2 - r1) * (c2 - c1):
                return Node(True, True, None, None, None, None)

            # 非叶子结点
            node1, node2, node3, node4 = (
                dfs(r1, c1, (r1 + r2) // 2, (c1 + c2) // 2),
                dfs(r1, (c1 + c2) // 2, (r1 + r2) // 2, c2),
                dfs((r1 + r2) // 2, c1, r2, (c1 + c2) // 2),
                dfs((r1 + r2) // 2, (c1 + c2) // 2, r2, c2),
            )
            return Node(False, False, node1, node2, node3, node4)

        preSumMatrix = PreSumMatrix(grid)

        # 左闭右开,递归时处理边界坐标会方便一些
        return dfs(0, 0, len(grid), len(grid[0]))

